feat!: collect telemetry from hypotheses updaters

tristanls opened the pull request:
This is an alternate proposal to implement the functionality in #396.

The pull request is marked `feat!` due to a breaking change, which changed the `EvidenceSlopeTracker.window_size` parameter default.

The overall design for collecting telemetry from hypotheses updaters is to update the protocol to return a tuple and then, in the case of `ResamplingHypothesesUpdater`, include telemetry in the returned tuple if configured to do so.

This proposal attempts to minimize coupling between code.

The `HypothesesUpdater` protocol update now requires that _any_ updater may return some telemetry, with the `DefaultHypothesesUpdater` returning `None`.

The `EvidenceGraphLM`, that uses hypotheses updaters, now collects any telemetry emitted by them.
```python
        if hypotheses_update_telemetry is not None:
            self.hypotheses_updater_telemetry[graph_id] = hypotheses_update_telemetry
```
This way, no matter what implementation of hypotheses updater is used, if it emits telemetry, it will be collected.

The learning module collects telemetry because of the coupling between Monty and MontyExperiment, where telemetry is collected via `_add_detailed_stats` call. Additionally, we _do not want_ the learning module to assume that it has a specific type of hypotheses updater, nor do we want it to reach into that one specific type of hypotheses updater and then reach into its tracker to do some computation.

I attempted to minimize changes to `TheoreticalLimitLMLoggingMixin`. The difference in implementation here vs #396 is that `TheoreticalLimitLMLoggingMixin` does not assume a specific telemetry format of the channel telemetry (aside from the generic dictionary of Any). It accepts channel telemetry and only appends what is in its scope to append. There is some weak coupling here in that it is technically possible for `TheoreticalLimitLMLoggingMixin` to override keys in channel telemetry, but addressing that is beyond the scope of this pull request.

Instead of reaching into the internals of `EvidenceSlopeTracker`, it now has `hyp_ages` method that can be called. Similarly, `_calculate_slopes` has been renamed to `calculate_slopes` as it is now part of its public API and called outside the class.

I thought that adding `compute_pose_errors` and reusing it in `compute_pose_error` was a better way to offer the quite different functionality, instead of using a parameter toggle.

Lastly, the use of the term "telemetry" is intentional as a marker for future refactoring. I am hoping that one day we'll end up emitting telemetry in a similar way we do `logger.info()` calls, out of band of the main code logic. When we get there, anything with "telemetry" in it will need to be refactored, and so this makes it easier to find (possibly maybe).

tristanls commented:
@ramyamounir, I asked for your review because this is the alternate proposal I mentioned. If you think this implements your prototype correctly, we can add others to the review.

ramyamounir reviewed (commented):
Really nice, thanks for taking the time to implement this change. 

I did consider this option of returning the logged statistics back to the parent LM vs. reaching into the composed functionality and extracting the logged statistics. I had two concerns with the former approach (one major and one minor):
- I assume that getting telemetry from a deeper composed class, such as the `HypothesesDisplacer`, would follow the same approach and return its telemetry to the Updater, and it gets included in the updater's telemetry (as a nested dictionary) then sent back to the parent LM. This works well if all the intermediate composed levels already follow the same protocol of sending back `Telemetry | None`. But if I have to implement this for the Displacer without it being already implemented in the Updater, I'd have to go implement it for the updater first, which may not sound much now, but it's definitely a lot of work for 4 or 5 levels of nested classes. This change would include updating the protocols and all the classes that follow these protocols for every level. To avoid this, we have to adopt this telemetry approach by default whenever extracting any functionality. I'm okay with this change btw.
- The current approach of sending back telemetry or None could break the hierarchical chain of telemetry. If I have `include_telemetry=False` for the Updater, but `include_telemetry=True` for the Displacer, the updater would not collect telemetry from it's child displacer even though it was defined in the configs. But this is probably an easy fix.

Note that I requested minor changes (e.g., unit tests or docstrings) in my review. Feel free to use the :eyes: to mean "do it yourself". I haven't tried it before, but I think maintainers are allowed to push commits to your PR branch if you've allowed it when creating the PR.

tristanls commented on src/tbp/monty/frameworks/models/evidence_matching/hypotheses_updater.py:
```diff
@@ -44,6 +44,9 @@
 
 logger = logging.getLogger(__name__)
 
+HypothesesUpdateTelemetry = Optional[Dict[str, Any]]
```
Correct, I tried to use `dict[str, Any] | None`, but using `|` in a type alias appears to require Python beyond 3.8 even with the `from __future__` import. So, I used `Optional` because other formulations did not work. Give it a try and see if you get an error, it's possible I made a mistake here.

ramyamounir replied:
Yes, you're right. It wants 3.10 or newer.

tristanls commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -275,10 +312,29 @@ def update_hypotheses(
             hypotheses_updates.append(channel_hypotheses)
 
+            if self.include_telemetry:
+                resampling_telemetry[input_channel] = asdict(
```
I initially did what you suggested, but then I noticed it requires the Mixin (and therefore the learning module) to know that the `channel_telemetry` has the `ChannelHypothesesResamplingTelemetry` type. This means that the Mixin knows the specific hypotheses updater class being used, which it is not allowed to know. If it knows, then we have coupling, and when we swap out an updater, things will break unexpectedly. It will be unexpected because our protocol says that anything that conforms to the protocol can be used.

So, I used the dataclass to ensure I didn't forget to populate all the fields. Still, I wanted a generic return value, i.e., `dict[str, Any]` so that the Mixin (and therefore the learning module) would not have to be coupled to the specific hypotheses updater implementation, but could work with the protocol and telemetry, which could be substituted for any other one. This desire to work with any telemetry format (as long as it is `dict[str, Any]`) is why Mixin doesn't assume anything beyond dictionary structure or attempt to read it, since it can't know the field names.

tristanls replied:
Another way to think about it, having `channel_telemetry: ChannelHypothesesResamplingTelemetry` would imply that the `HypothesesUpdater(Protocol)` be:
```python
class HypothesesUpdater(Protocol):
    def update_hypotheses(
        self,
        hypotheses: Hypotheses,
        features: dict,
        displacements: dict | None,
        graph_id: str,
        mapper: ChannelMapper,
        evidence_update_threshold: float,
    ) -> tuple[list[ChannelHypotheses], ChannelHypothesesResamplingTelemetry]:
        """Update hypotheses based on sensor displacement and sensed features.
```
See the problem?

ramyamounir replied:
That all makes sense now, it's because we're calling the `_channel_resampling_telemetry` for all updater types. We should probably change the function name and the stats key in the mixin to be something that doesn't mention resampling then. Maybe:

```python
        ...
        stats["hypotheses_updater_telemetry"] = self._hypotheses_updater_telemetry()
        return stats
```

One more nitpick: I ran it with `include_telemetry=False` and it stored a list of empty dictionaries for each episode - as expected. It's a nitpick but i'm not sure if we want to completely disable the logging or allow the empty dictionaries to be written. They don't take much space.

tristanls replied:
> Does the type checker still complain if we use Any for the protocol and a more specific type for the classes?

It should. This violates the Liskov Substitution Principle, which states that a protocol instance should be replaceable by any of its implementations.

`DefaultHypothesesUpdater` returning `tuple[list[ChannelHypotheses], None]` is fine, because `None` is allowed by the protocol.

`ResamplingHypothesesUpdater` is _**not substitutable**_ for the protocol because  `dict[str, ChannelHypothesesResamplingTelemetry]` is _**more specific**_ than `dict[str, Any] | None`.

The concrete protocol implementation can have a _less specific_ type, but not a more restrictive type.

Simpler example:
```python
from __future__ import annotations

from typing import Any, Protocol


class Something(Protocol):
    def accept(self, num: int | str) -> None: ...


class MyInt:
    def accept(self, num: int) -> None:
        print(f"MyInt: {num}")


class MyFloat:
    def accept(self, num: str) -> None:
        print(f"MyFloat: {num}")

class MyAny:
    def accept(self, num: Any) -> None:
        print(f"MyAny: {num}")

def check(acceptor: Something) -> None:
    pass

check(MyInt())  # fails type check
check(MyFloat())  # fails type check
check(MyAny()) # OK, can handle int, or str, or anything else
```

tristanls commented on src/tbp/monty/frameworks/models/evidence_matching/resampling_hypotheses_updater.py:
```diff
@@ -48,6 +51,24 @@
+        TODO: Additional description here regarding availability of hypotheses
+        identified by `removed_ids`.
+    """
+
```
I think I have an idea 💡 ...

tristanls replied:
My idea aside... I think I remember now why I did this.

Whatever telemetry `update_hypotheses` returns, we want it grouped by the input channel. Returning a list would work, but in that case, we would have to specify a protocol for the telemetry that we could use to interrogate what input channel the telemetry corresponds to:
```python
class InputChannelTelemetry(Protocol):
    @property
    def input_channel(self) -> str: ...
```
Then, we would have (gotta remember that `Optional`):
```python
HypothesesUpdateTelemetry = Optional[list[InputChannelTelemetry]]
```
The return value from `update_hypotheses` would still be:
```python
-> tuple[list[ChannelHypotheses], HypothesesUpdateTelemetry]
```
Then, `_resampling_telemetry` takes the shape of:
```python
    def _resampling_telemetry(self) -> dict[str, dict[str, dict[str, Any]]]:
        """Compile resampling telemetry across all objects and input channels.

        Returns:
            A nested dictionary of {graph_id: {input_channel: resampling_telemetry}}.
        """
        stats = {}
        for graph_id, graph_telemetry in self.hypotheses_updater_telemetry.items():
            stats[graph_id] = {
                channel_telemetry.input_channel: self._channel_resampling_telemetry(
                    graph_id, channel_telemetry
                )
                for channel_telemetry in graph_telemetry
            }
        return stats
```

It's pretty much equivalent in functionality. However, using a `dict[str, Any]` and declaring that the keys are input channels retains the same amount of coupling but lets us avoid writing the `InputChannelTelemetry` protocol. Maybe being explicit with the protocol is good.

ramyamounir replied:
I was planning to just pop the input channel after we convert it from a `ChannelHypothesesResamplingTelemetry` to a `dict`. Even if this works, it's not better than what you have already implemented. I'm just commenting to check my understanding. What you have already looks great to me :+1: 

```python
    def _resampling_telemetry(self) -> dict[str, dict[str, dict[str, Any]]]:
        """Compile resampling telemetry across all objects and input channels.

        Returns:
            A nested dictionary of {graph_id: {input_channel: resampling_telemetry}}.
        """
        stats = {}
        for graph_id, graph_telemetry in self.hypotheses_updater_telemetry.items():
            stats[graph_id] = {
                channel_telemetry.input_channel: self._channel_resampling_telemetry(
                    graph_id, channel_telemetry
                )
                for channel_telemetry in graph_telemetry
            }
        return stats

    def _channel_resampling_telemetry(
        self, graph_id: str, channel_telemetry: dict[str, Any]
    ) -> dict[str, Any]:
        """Assemble resampling telemetry for specific graph ID and input channel.

        Args:
            graph_id: The graph ID.
            channel_telemetry: Telemetry for the specific input channel.

        Returns:
            A dictionary containing telemetry in `channel_telemetry` and:
                - evidence: The hypotheses evidence scores.
                - rotations: Rotations of the hypotheses. Note that the buffer encoder
                    will encode those as euler "xyz" rotations in degrees.
                - pose_errors: Rotation errors relative to the target pose.
        """
        stats: dict[str, Any] = asdict(channel_telemetry).copy()
        input_channel = stats.pop("input_channel")

        mapper = self.channel_hypothesis_mapping[graph_id]
        channel_rotations = mapper.extract(self.possible_poses[graph_id], input_channel)
        channel_rotations_inv = Rotation.from_matrix(channel_rotations).inv()
        channel_evidence = mapper.extract(self.evidence[graph_id], input_channel)

        stats["evidence"] = channel_evidence
        stats["rotations"] = channel_rotations_inv
        stats["pose_errors"] = compute_pose_errors(
            channel_rotations_inv,
            Rotation.from_quat(self.primary_target_rotation_quat),
        )
        return stats

```

tristanls replied:
> ```python
>     def _channel_resampling_telemetry(
>         self, graph_id: str, channel_telemetry: dict[str, Any]
>     ) -> dict[str, Any]:
>         stats: dict[str, Any] = asdict(channel_telemetry).copy()
>         input_channel = stats.pop("input_channel")
>         ...
> ```
This is not the appropriate place for this operation. Note that `stats: dict[str, Any]`. There is no reason (from this code's perspective and the information passed into it) to assume at this point that `"input_channel"` is a valid key here. This is an information leak from a detailed implementation (`ChannelHypothesesResamplingTelemetry`) into a more abstract general code that should not have this detailed information (`dict[str, Any]`). If you wanted to pop `"input_channel"`, it should be done in the code that knows that key is there, i.e., the `ResamplingHypothesesUpdater` that constructs `ChannelHypothesesResamplingTelemetry`. But that's what I did by not including it in the first place.

This is also not a great example, because as I mentioned before, there is weak coupling in that we assume we can add `"evidence"`, `"rotations"`, and `"pose_errors"` without overriding what's in `channel_telemetry`. It might actually be better if we namespace things like so:
```python
        stats: dict[str, Any] = {}
        stats["channel_telemetry"] = channel_telemetry.copy()
        stats["evidence"] = channel_evidence
        stats["rotations"] = channel_rotations_inv
        stats["pose_errors"] = compute_pose_errors(
            channel_rotations_inv,
            Rotation.from_quat(self.primary_target_rotation_quat),
        )
        return stats
```
The above change is in https://github.com/thousandbrainsproject/tbp.monty/pull/400/commits/6010e84022a793526a98dd9f3f52bd4a7d6455cb

jeremyshoemaker replied:
It bothers me how much we use dictionaries, especially nested ones all over the place. It makes the typing much harder to do. Improving it here is probably out of scope though.

tristanls replied:
@jeremyshoemaker, updated: https://github.com/thousandbrainsproject/tbp.monty/pull/400/commits/0b45e8a0a812f6de93b845b190b9793bc248b713

ramyamounir replied:
One minor issue here is that we need to serialize `HypothesesUpdaterChannelTelemetry` before we add it to the dictionary for logging.

```python
        for graph_id, graph_telemetry in self.hypotheses_updater_telemetry.items():
            stats[graph_id] = {
                input_channel: self._channel_telemetry(
                    graph_id, input_channel, channel_telemetry
                ) # returns `HypothesesUpdaterChannelTelemetry
                for input_channel, channel_telemetry in graph_telemetry.items()
            }
```

ramyamounir replied:
Looks and works great, thanks.

tristanls commented:
@ramyamounir, requesting another review. I believe I addressed your comments. Please note that in the latest batch of changes, I nested the hypotheses updater telemetry one level deeper in the `stats` object. This will impact the use of the data for visualization. Reasons are in the comment thread above.
```python
        stats: dict[str, Any] = {}
        stats["hypotheses_updater"] = channel_telemetry.copy()  # <- extra nesting/namespacing here
        stats["evidence"] = channel_evidence
        stats["rotations"] = channel_rotations_inv
        stats["pose_errors"] = compute_pose_errors(
            channel_rotations_inv,
            Rotation.from_quat(self.primary_target_rotation_quat),
        )
        return stats
```

tristanls commented on src/tbp/monty/frameworks/models/mixins/no_reset_evidence.py:
```diff
@@ -7,14 +7,39 @@
+    hypotheses_updater: dict[str, Any]
+    evidence: npt.NDArray[np.float64]
+    rotations: npt.NDArray[np.float64]
+    pose_errors: npt.NDArray[np.float64] | float
```
@ramyamounir, I think this could be fine without the `float`, but I want to verify with you.

ramyamounir replied:
Yes, I see the cast you added. Looks good!

ramyamounir commented:
Thanks for all these edits @tristanls, there is just one minor unresolved issue. I don't have any other concerns apart from the high-level bullet points in my original comment, but this is something we can discuss in one of the meetings.

tristanls commented:
I pressed the button too quickly 😕... I should have added a Co-Authored By trailer to indicate that you're the original author @ramyamounir. My apologies.
